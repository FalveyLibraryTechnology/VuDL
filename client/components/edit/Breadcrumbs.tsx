import styles from "../shared/Breadcrumbs.module.css";
import React, { useEffect, useState } from "react";
import BasicBreadcrumbs from "../shared/BasicBreadcrumbs";
import { TreeNode, processBreadcrumbData } from "../../util/Breadcrumbs";
import { useEditorContext } from "../../context/EditorContext";
import Link from "next/link";

interface BreadcrumbsProps {
    pid?: string | null;
}

const Breadcrumbs = ({ pid = null }: BreadcrumbsProps): React.ReactElement => {
    const {
        state: { parentDetailsStorage, topLevelPids },
        action: { loadParentDetailsIntoStorage },
    } = useEditorContext();
    const [shallow, setShallow] = useState<boolean>(true);

    const dataForPid =
        pid !== null && Object.prototype.hasOwnProperty.call(parentDetailsStorage, pid as string)
            ? parentDetailsStorage[pid]
            : {};
    const key = shallow ? "shallow" : "full";
    const loaded = Object.prototype.hasOwnProperty.call(dataForPid, key);

    useEffect(() => {
        if (!loaded && pid !== null) {
            loadParentDetailsIntoStorage(pid, shallow);
        }
    }, [loaded, shallow]);

    // Special case: no PID, we're at the top level:
    if (pid === null) {
        return <BasicBreadcrumbs />;
    }

    if (!loaded) {
        return <span>Loading...</span>;
    }
    const treeData: Array<Array<TreeNode>> = processBreadcrumbData(parentDetailsStorage[pid][key]).paths;

    const contents = treeData.map((trail, trailIndex: number) => {
        const keySuffix = trailIndex + "_" + (shallow ? "s" : "f");
        const trailPids: Array<string> = [];
        const breadcrumbs = trail.map((breadcrumb) => {
            trailPids.push(breadcrumb.pid);
            return (
                <li key={"breadcrumb_" + breadcrumb.pid + "_" + keySuffix}>
                    <Link href={"/edit/object/" + breadcrumb.pid}>{breadcrumb.title}</Link>
                </li>
            );
        });
        // If we're in shallow mode, and our trail is non-empty and does not include the uppermost
        // top-level PID, we should show an expand control.
        if (shallow && trailPids.length > 0 && !trailPids.includes(topLevelPids[0] ?? "")) {
            breadcrumbs.unshift(
                <li key={"breadcrumb_expand_" + keySuffix}>
                    <button onClick={() => setShallow(false)}>...</button>
                </li>
            );
        }
        breadcrumbs.unshift(
            <li key={"breadcrumb_home_" + keySuffix}>
                <Link href="/edit">Edit Home</Link>
            </li>
        );
        breadcrumbs.unshift(
            <li key={"breadcrumb_mainmenu_" + keySuffix}>
                <Link href="/">Main Menu</Link>
            </li>
        );
        return (
            <ul className={styles.breadcrumb} key={"breadcrumbs_" + keySuffix}>
                {breadcrumbs}
            </ul>
        );
    });
    return <>{contents}</>;
};

export default Breadcrumbs;
