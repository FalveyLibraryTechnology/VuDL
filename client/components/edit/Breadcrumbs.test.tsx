import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { waitFor } from "@testing-library/react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import Breadcrumbs from "./Breadcrumbs";
import { EditorContextProvider } from "../../context/EditorContext";
import { FetchContextProvider } from "../../context/FetchContext";

describe("Breadcrumb", () => {
    let props: Record<string, string>;
    let lastRequestUrl: string;
    let breadcrumbResponse = {};

    beforeEach(() => {
        props = { pid: "foo:1234" };
        global.fetch = jest.fn((url) => {
            lastRequestUrl = url;
            return {
                ok: true,
                status: 200,
                json: async function () {
                    return breadcrumbResponse;
                },
            };
        });
    });

    async function runStandardSnapshotTest(initiallyShallow: boolean) {
        const wrapper = mount(
            <FetchContextProvider>
                <EditorContextProvider>
                    <Breadcrumbs {...{...props, initiallyShallow}} />
                </EditorContextProvider>
            </FetchContextProvider>
        );
        await waitFor(() => expect(global.fetch).toHaveBeenCalledTimes(1));
        expect(lastRequestUrl).toEqual("http://localhost:9000/api/edit/object/foo%3A1234/parents" + (initiallyShallow ? "?shallow=1" : ""));
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    }

    function getObject(pid, title, parents = []) {
        return { pid, title, parents };
    }

    it("renders without a pid", () => {
        props = {};
        const wrapper = mount(
            <FetchContextProvider>
                <EditorContextProvider>
                    <Breadcrumbs />
                </EditorContextProvider>
            </FetchContextProvider>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders using ajax-loaded object data (no parents, shallow mode)", async () => {
        breadcrumbResponse = getObject("foo:1234", "Fake Title");
        await runStandardSnapshotTest(true);
    });

    it("renders using ajax-loaded object data (no parents, deep mode)", async () => {
        breadcrumbResponse = getObject("foo:1234", "Fake Title");
        await runStandardSnapshotTest(false);
    });

    it("renders using ajax-loaded object data (simple case, shallow mode)", async () => {
        const parent = getObject("foo:1233", "Fake Parent");
        breadcrumbResponse = getObject("foo:1234", "Fake Title", [parent]);
        await runStandardSnapshotTest(true);
    });

    it("renders using ajax-loaded object data (simple case, deep mode)", async () => {
        const parent = getObject("foo:1233", "Fake Parent");
        breadcrumbResponse = getObject("foo:1234", "Fake Title", [parent]);
        await runStandardSnapshotTest(false);
    });

    it("renders using ajax-loaded object data (many parents, shallow mode)", async () => {
        const parents = [];
        for (let $x = 0; $x < 15; $x++) {
            parents.push(getObject("foo:" + $x, "Fake Parent " + $x));
        }
        breadcrumbResponse = getObject("foo:1234", "Fake Title", parents);
        await runStandardSnapshotTest(true);
    });

    it("renders using ajax-loaded object data (many parents, deep mode)", async () => {
        const parents = [];
        for (let $x = 0; $x < 15; $x++) {
            parents.push(getObject("foo:" + $x, "Fake Parent " + $x));
        }
        breadcrumbResponse = getObject("foo:1234", "Fake Title", parents);
        await runStandardSnapshotTest(false);
    });

    it("renders using ajax-loaded object data (deep parent chain)", async () => {
        let parents = [];
        for (let $x = 0; $x < 15; $x++) {
            parents = [getObject("foo:" + $x, "Fake Parent " + $x, parents)];
        }
        breadcrumbResponse = getObject("foo:1234", "Fake Title", parents);
        await runStandardSnapshotTest(false);
    });

    it("renders using ajax-loaded object data (multiple parents with common grandparent)", async () => {
        const grandparent = getObject("foo:1232", "Fake Grandparent");
        const parent1 = getObject("foo:1233", "Fake Parent 1", [grandparent]);
        const parent2 = getObject("foo:1231", "Fake Parent 2", [grandparent]);
        breadcrumbResponse = getObject("foo:1234", "Fake Title", [parent1, parent2]);
        await runStandardSnapshotTest(false);
    });
});
