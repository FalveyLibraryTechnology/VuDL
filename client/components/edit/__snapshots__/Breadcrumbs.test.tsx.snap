// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb renders using ajax-loaded object data (deep parent chain) 1`] = `
<FetchContextProvider>
  <EditorContextProvider>
    <Breadcrumbs
      pid="foo:1234"
    >
      <ul
        className="breadcrumb"
        key="breadcrumbs_0_s"
      >
        <li
          key="breadcrumb_mainmenu_0_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_0_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_0_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:0_0_s"
        >
          <Link
            href="/edit/object/foo:0"
          >
            <a
              href="/edit/object/foo:0"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 0
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:1_0_s"
        >
          <Link
            href="/edit/object/foo:1"
          >
            <a
              href="/edit/object/foo:1"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 1
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:2_0_s"
        >
          <Link
            href="/edit/object/foo:2"
          >
            <a
              href="/edit/object/foo:2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 2
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:3_0_s"
        >
          <Link
            href="/edit/object/foo:3"
          >
            <a
              href="/edit/object/foo:3"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 3
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:4_0_s"
        >
          <Link
            href="/edit/object/foo:4"
          >
            <a
              href="/edit/object/foo:4"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 4
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:5_0_s"
        >
          <Link
            href="/edit/object/foo:5"
          >
            <a
              href="/edit/object/foo:5"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 5
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:6_0_s"
        >
          <Link
            href="/edit/object/foo:6"
          >
            <a
              href="/edit/object/foo:6"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 6
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:7_0_s"
        >
          <Link
            href="/edit/object/foo:7"
          >
            <a
              href="/edit/object/foo:7"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 7
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:8_0_s"
        >
          <Link
            href="/edit/object/foo:8"
          >
            <a
              href="/edit/object/foo:8"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 8
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:9_0_s"
        >
          <Link
            href="/edit/object/foo:9"
          >
            <a
              href="/edit/object/foo:9"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 9
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:10_0_s"
        >
          <Link
            href="/edit/object/foo:10"
          >
            <a
              href="/edit/object/foo:10"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 10
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:11_0_s"
        >
          <Link
            href="/edit/object/foo:11"
          >
            <a
              href="/edit/object/foo:11"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 11
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:12_0_s"
        >
          <Link
            href="/edit/object/foo:12"
          >
            <a
              href="/edit/object/foo:12"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 12
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:13_0_s"
        >
          <Link
            href="/edit/object/foo:13"
          >
            <a
              href="/edit/object/foo:13"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 13
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:14_0_s"
        >
          <Link
            href="/edit/object/foo:14"
          >
            <a
              href="/edit/object/foo:14"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 14
            </a>
          </Link>
        </li>
      </ul>
    </Breadcrumbs>
  </EditorContextProvider>
</FetchContextProvider>
`;

exports[`Breadcrumb renders using ajax-loaded object data (many parents) 1`] = `
<FetchContextProvider>
  <EditorContextProvider>
    <Breadcrumbs
      pid="foo:1234"
    >
      <ul
        className="breadcrumb"
        key="breadcrumbs_0_s"
      >
        <li
          key="breadcrumb_mainmenu_0_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_0_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_0_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:0_0_s"
        >
          <Link
            href="/edit/object/foo:0"
          >
            <a
              href="/edit/object/foo:0"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 0
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_1_s"
      >
        <li
          key="breadcrumb_mainmenu_1_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_1_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_1_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:1_1_s"
        >
          <Link
            href="/edit/object/foo:1"
          >
            <a
              href="/edit/object/foo:1"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 1
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_2_s"
      >
        <li
          key="breadcrumb_mainmenu_2_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_2_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_2_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:2_2_s"
        >
          <Link
            href="/edit/object/foo:2"
          >
            <a
              href="/edit/object/foo:2"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 2
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_3_s"
      >
        <li
          key="breadcrumb_mainmenu_3_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_3_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_3_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:3_3_s"
        >
          <Link
            href="/edit/object/foo:3"
          >
            <a
              href="/edit/object/foo:3"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 3
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_4_s"
      >
        <li
          key="breadcrumb_mainmenu_4_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_4_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_4_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:4_4_s"
        >
          <Link
            href="/edit/object/foo:4"
          >
            <a
              href="/edit/object/foo:4"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 4
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_5_s"
      >
        <li
          key="breadcrumb_mainmenu_5_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_5_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_5_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:5_5_s"
        >
          <Link
            href="/edit/object/foo:5"
          >
            <a
              href="/edit/object/foo:5"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 5
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_6_s"
      >
        <li
          key="breadcrumb_mainmenu_6_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_6_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_6_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:6_6_s"
        >
          <Link
            href="/edit/object/foo:6"
          >
            <a
              href="/edit/object/foo:6"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 6
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_7_s"
      >
        <li
          key="breadcrumb_mainmenu_7_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_7_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_7_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:7_7_s"
        >
          <Link
            href="/edit/object/foo:7"
          >
            <a
              href="/edit/object/foo:7"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 7
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_8_s"
      >
        <li
          key="breadcrumb_mainmenu_8_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_8_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_8_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:8_8_s"
        >
          <Link
            href="/edit/object/foo:8"
          >
            <a
              href="/edit/object/foo:8"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 8
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_9_s"
      >
        <li
          key="breadcrumb_mainmenu_9_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_9_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_9_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:9_9_s"
        >
          <Link
            href="/edit/object/foo:9"
          >
            <a
              href="/edit/object/foo:9"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 9
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_10_s"
      >
        <li
          key="breadcrumb_mainmenu_10_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_10_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_10_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:10_10_s"
        >
          <Link
            href="/edit/object/foo:10"
          >
            <a
              href="/edit/object/foo:10"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 10
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_11_s"
      >
        <li
          key="breadcrumb_mainmenu_11_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_11_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_11_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:11_11_s"
        >
          <Link
            href="/edit/object/foo:11"
          >
            <a
              href="/edit/object/foo:11"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 11
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_12_s"
      >
        <li
          key="breadcrumb_mainmenu_12_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_12_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_12_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:12_12_s"
        >
          <Link
            href="/edit/object/foo:12"
          >
            <a
              href="/edit/object/foo:12"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 12
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_13_s"
      >
        <li
          key="breadcrumb_mainmenu_13_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_13_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_13_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:13_13_s"
        >
          <Link
            href="/edit/object/foo:13"
          >
            <a
              href="/edit/object/foo:13"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 13
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_14_s"
      >
        <li
          key="breadcrumb_mainmenu_14_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_14_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_14_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:14_14_s"
        >
          <Link
            href="/edit/object/foo:14"
          >
            <a
              href="/edit/object/foo:14"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 14
            </a>
          </Link>
        </li>
      </ul>
    </Breadcrumbs>
  </EditorContextProvider>
</FetchContextProvider>
`;

exports[`Breadcrumb renders using ajax-loaded object data (multiple parents with common grandparent) 1`] = `
<FetchContextProvider>
  <EditorContextProvider>
    <Breadcrumbs
      pid="foo:1234"
    >
      <ul
        className="breadcrumb"
        key="breadcrumbs_0_s"
      >
        <li
          key="breadcrumb_mainmenu_0_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_0_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_0_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:1232_0_s"
        >
          <Link
            href="/edit/object/foo:1232"
          >
            <a
              href="/edit/object/foo:1232"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Grandparent
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:1233_0_s"
        >
          <Link
            href="/edit/object/foo:1233"
          >
            <a
              href="/edit/object/foo:1233"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 1
            </a>
          </Link>
        </li>
      </ul>
      <ul
        className="breadcrumb"
        key="breadcrumbs_1_s"
      >
        <li
          key="breadcrumb_mainmenu_1_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_1_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_1_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:1232_1_s"
        >
          <Link
            href="/edit/object/foo:1232"
          >
            <a
              href="/edit/object/foo:1232"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Grandparent
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_foo:1231_1_s"
        >
          <Link
            href="/edit/object/foo:1231"
          >
            <a
              href="/edit/object/foo:1231"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent 2
            </a>
          </Link>
        </li>
      </ul>
    </Breadcrumbs>
  </EditorContextProvider>
</FetchContextProvider>
`;

exports[`Breadcrumb renders using ajax-loaded object data (no parents) 1`] = `
<FetchContextProvider>
  <EditorContextProvider>
    <Breadcrumbs
      pid="foo:1234"
    >
      <ul
        className="breadcrumb"
        key="breadcrumbs_0_s"
      >
        <li
          key="breadcrumb_mainmenu_0_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_0_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
      </ul>
    </Breadcrumbs>
  </EditorContextProvider>
</FetchContextProvider>
`;

exports[`Breadcrumb renders using ajax-loaded object data (simple case) 1`] = `
<FetchContextProvider>
  <EditorContextProvider>
    <Breadcrumbs
      pid="foo:1234"
    >
      <ul
        className="breadcrumb"
        key="breadcrumbs_0_s"
      >
        <li
          key="breadcrumb_mainmenu_0_s"
        >
          <Link
            href="/"
          >
            <a
              href="/"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Main Menu
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_home_0_s"
        >
          <Link
            href="/edit"
          >
            <a
              href="/edit"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Edit Home
            </a>
          </Link>
        </li>
        <li
          key="breadcrumb_expand_0_s"
        >
          <button
            onClick={[Function]}
          >
            ...
          </button>
        </li>
        <li
          key="breadcrumb_foo:1233_0_s"
        >
          <Link
            href="/edit/object/foo:1233"
          >
            <a
              href="/edit/object/foo:1233"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Fake Parent
            </a>
          </Link>
        </li>
      </ul>
    </Breadcrumbs>
  </EditorContextProvider>
</FetchContextProvider>
`;

exports[`Breadcrumb renders without a pid 1`] = `
<FetchContextProvider>
  <EditorContextProvider>
    <Breadcrumbs>
      <BasicBreadcrumbs>
        <ul
          className="breadcrumb"
        >
          <li>
            <Link
              href="/"
            >
              <a
                href="/"
                onClick={[Function]}
                onMouseEnter={[Function]}
              >
                Main Menu
              </a>
            </Link>
          </li>
        </ul>
      </BasicBreadcrumbs>
    </Breadcrumbs>
  </EditorContextProvider>
</FetchContextProvider>
`;
